8

LiveNathan opened the pull request:
Several of the APIs were either dead or changed, so I made some modifications.

LiveNathan committed:
Refactor HTML, CSS, JS files and add new module

Refactored the JS, HTML, and CSS files in tic-tac-toe, portfolio, triangle solver and added a new module in API requests (geo-location) application.

The changes were made to improve overall readability, semantics and performance of the codes. HTML files in portfolio and tic-tac-toe were refactored for more semantic layout using main, div, article, and aside.

JS file in triangle solver was refactored to improve performance by adding event listener to parent of calculator inputs rather than each input individually. CSS in portfolio was refactored to improve the design.

New module was added in API requests to fetch and display geo-location information of the user.

LiveNathan committed:
Add additional geographical information to display

Enhanced the geographical information displayed in `08-09-apis_requests/index.html` by including the country code, country flag as an emoji, and other specifics. This will provide a more comprehensive understanding of the user's current location. The data is fetched from the ipstack API and displayed on the webpage.

LiveNathan committed:
"Update with additional location and time details on index page"

In the updated version of `index.html`, a few more details related to the user's location and local time have been added. These include the area, languages, and regional bloc information of the user's country. There is now a correct spelling of the word 'Population'.

Additions to display local sunrise and sunset times, the length of the day, and the time of solar noon have also been incorporated using a new API(`https://sunrise-sunset.org/api`). More information is now fetched from existing APIs to enrich the user's location and time details.

The changes aim to provide users with more comprehensive information about their current location.

LiveNathan committed:
"Enhance error handling and improve background color transition"

This update enhances error handling in getIpStack, getWorldTime, getCountries, and getSun functions with the inclusion of try/catch blocks. This allows better debug logging and prevents the application from breaking in case of errors in API calls.

Also, a background color transition that smoothly changes according to sunrise, noon, sunset, and midnight was added. This feature takes timezone into account by using data from the sundata and worldTime API. This creates a more interactive user experience.

LiveNathan committed:
Add SVG files for sun and moon and improve UI

This commit adds SVG files for a sun and moon icon, utilized for a more visual representation of sunrise and sunset times in the application. Changes were made in the index.html where the sun.svg and moon.svg files were referenced.

The CSS was updated to display the sun and moon icons appropriately and JavaScript was modified to control their visibility based on sunrise or sunset times. The commit also improved the UI by changing the background colour and text colour dynamically based on the time of day, enhancing user experience with regard to sunrise and sunset times in specific geographical areas.

The modifications also restructured the API request to parse dates in ISO format, so the conversion of string to time logic is no longer necessary. Remaining time calculation is handled by a helper function to ensure code organization.

This approach will make the UI more intuitive and appealing to users.

LiveNathan committed:
Add CSS and JS files for ipStack, worldTime and sunrise APIs

Added two separate files, default.css and index.js, for handling styles and scripts respectively. This change was made to keep the HTML file clean and improve readability. The styles code, previously embedded into HTML file, is now moved to default.css file. The scripts code is moved to index.js file. This code includes multiple async calls to various APIs such as ipStack, worldTime and sunrise. These API calls fetch location details, current world time and sunrise-sunset time of the client's location, and dynamically update the HTML elements using the fetched data.

iansedano commented on 08-09-apis_requests/index.js:
```diff
@@ -0,0 +1,180 @@
+const IPSTACK_ACCESS_KEY = "315c943e1895202bbf4fc8129849e4b4";
+const ipStackPromise = getIpStack();
+const worldTimePromise = getWorldTime();
+const sunPromise = getSun();
```
Using top level promises in this way can lead to unexpected behaviors and makes things hard to maintain.

Better to call the promises from within the functions that need them.

If you need the results from one promise in multiple places. Then you can have a wrapping function that calls the promise, packages the information in a data structure, and then distributes it where it needs to go.

For example:

```js
async function main(){
	const ipResponse = await getIpStack(); // only makes request and returns response
	const json = await ipResponse.json();
	const ipData = mapIpData(json) // optional: process json into simpler and more focussed format (not async)
	renderIpData(ipData); // updates the page with ip data (not async)

	const worldTimeResponse = await getWorldTime(ipData); // <--passing in ipData
	// ... etc
}
```

Note: this is just something off the top of my head. It may not be the best way to organize the code at a high level.

iansedano commented on 08-09-apis_requests/index.js:
```diff
@@ -0,0 +1,180 @@
+getIssPeople();
+
+async function getSunMultiplier() {
+    const {sunriseTime, sunsetTime, solarNoonTime} = await sunPromise;
```
Nice destructuring!

LiveNathan replied:
destructuring? You think you can just go around making up words?

iansedano reviewed (changes requested):
Hi Nathan, great work!

I've left just a couple comments for now. They are quite high level suggestions that will affect much of the code, so I didn't want to get too nitty gritty. That said, there isn't much nitty gritty to pick on, anyway -- the code is well written!

The only thing that seems to not work well for me is the position of the sun and moon when rendered on the page.

In terms of the high level organization. I'd expect for most of the functions to be broken up so that each of their responsabilities are well defined. Especially when it comes to `fetch`. Generally the pattern that you use is:

- First write functions to request the data where you can handle all the errors that are typical to fetching.
- Then write functions to process/tidy your data.
- Then write functions to render your data.

As a final example. The sun position: you'll already have a function that just fetches world time data. Then have a function or functions that process the world time data and returnsa number that represents the position of the sun/moon. Then a function that takes this number and renders it on the page as the sun/moon.

Currently, the data processing is spread over two functions. First you have some logic to calculate the multiplier, that could be in it's own function. Then you have some logic to translate the multiplier into a position/angle -- which could also be it's own function.

In any case, great job!

LiveNathan reviewed (commented):
done

LiveNathan committed:
Refactor variable scope and async function calls

The changes made brought the IPSTACK_ACCESS_KEY constant into the getIpStack function where it is being used. Every occurrence of promises (ipStackPromise, worldTimePromise, sunPromise) directly used in the code were replaced with the async functions they were assigned to. This is done to ensure encapsulation of the IPSTACK_ACCESS_KEY constant and to simplify the structure, utilizing async/await more directly.
